release/nightly_tests/dataset/multi_node_train_benchmark.py

```diff
@@ -98,81 +146,212 @@ def get_transform(to_torch_tensor):
     return transform
 
 
-def crop_and_flip_image_batch(image_batch):
+def crop_and_flip_image(row):
     transform = get_transform(False)
-    batch_size, height, width, channels = image_batch["image"].shape
-    tensor_shape = (batch_size, channels, height, width)
-    image_batch["image"] = transform(
-        torch.Tensor(image_batch["image"].reshape(tensor_shape))
-    )
-    return image_batch
+    # Make sure to use torch.tensor here to avoid a copy from numpy.
+    row["image"] = transform(torch.tensor(np.transpose(row["image"], axes=(2, 0, 1))))
+    return row
 
 
-if __name__ == "__main__":
-    args = parse_args()
-    metrics = {}
+def benchmark_code(
+    args,
+    cache_output_ds=False,
+    cache_input_ds=False,
+    prepartition_ds=False,
+):
+    """
+    - cache_output_ds: Cache output dataset (ds.materialize()) after preprocessing fn.
+    - cache_input_ds: Cache input dataset, then apply a preprocessing fn.
+    - prepartition_ds: Pre-partition and cache input dataset across workers.
+    """
+    assert (
+        sum([cache_output_ds, cache_input_ds, prepartition_ds]) <= 1
+    ), "Can only test one caching variant at a time"
 
     # 1) Read in data with read_images() / read_parquet()
     if args.file_type == "image":
-        ray_dataset = ray.data.read_images(args.data_root)
+        ray_dataset = ray.data.read_images(
+            args.data_root,
+            mode="RGB",
+        )
     elif args.file_type == "parquet":
-        ray_dataset = ray.data.read_parquet(args.data_root)
+        ray_dataset = ray.data.read_parquet(
+            args.data_root,
+        )
     else:
         raise Exception(f"Unknown file type {args.file_type}")
-    # 2) Preprocess data by applying transformation with map_batches()
-    ray_dataset = ray_dataset.map_batches(crop_and_flip_image_batch)
+
+    if cache_input_ds:
+        ray_dataset = ray_dataset.materialize()
+
+    # 2) Preprocess data by applying transformation with map/map_batches()
+    ray_dataset = ray_dataset.map(crop_and_flip_image)
+    if cache_output_ds:
+        ray_dataset = ray_dataset.materialize()
 
     def train_loop_per_worker():
         it = train.get_dataset_shard("train")
+        device = train.torch.get_device()
 
         for i in range(args.num_epochs):
+            print(f"Epoch {i+1} of {args.num_epochs}")
             num_rows = 0
             start_t = time.time()
-            for batch in it.iter_batches(
+            for batch in it.iter_torch_batches(
                 batch_size=args.batch_size,
-                local_shuffle_buffer_size=args.local_shuffle_buffer_size,
-                prefetch_batches=10,
             ):
                 num_rows += args.batch_size
             end_t = time.time()
-            # Record throughput per epoch.
-            epoch_tput = num_rows / (end_t - start_t)
-            train.report({"tput": epoch_tput, "epoch": i})
+        # Workaround to report the final epoch time from each worker, so that we
```

Comment:
nit: move this loop function out of the main function? as it's getting too long.